### ISSUE TITLE
 KeyBinding Issue when attaching it to windows with commands with canexcute methods

### ISSUE THREAD

**Issue by ahmedflix25** [Caliburn-Micro/Caliburn.Micro#657]
if i add for example `cal:Message.Attach="[Key Enter] = [EnterPressed]"` to the window/page/panel etc and the `EnterPressed` command had a false CanExcute method it will set `IsEnable` property for that controller to false and that will disable the window/page/panel and every element in it and if that KeyBinding have more that one combination and they all true but one it will also do the same thing

an example project 
https://github.com/ahmedflix25/CaliburnMicroKeyBinding

if the CanExcute method is false it should only disable that KeyBinding and nothing else

**Comment by Filipsi** [comment]
I can confirm this problem.

When guard method prevents execution, Caliburn disables the control associated with the action.
In this case it disables whatever control has the [KeyTrigger](https://github.com/Caliburn-Micro/Caliburn.Micro/blob/f435e3c21ec20a56b905ff50a8a1eae3c1b7789c/samples/scenarios/Scenario.KeyBinding/Input/KeyTrigger.cs), since it technically is target of the action.
Even tho that in our case, the control is only used to capture KeyDown events.

**Comment by ahmedflix25** [comment]
a work around would be something like this

`void KeyTriggerForStuff()
{
    if (CanDoStuff)
    {
        DoStuff();
    }
}`

and you just bind the trigger to that new method

**Comment by Filipsi** [comment]
Possible solution might be addition of new dependency property to [ActionMessage](https://github.com/Caliburn-Micro/Caliburn.Micro/blob/63a49070f765c126d73171722b30f4ca7471ab6d/src/Caliburn.Micro.Platform/ActionMessage.cs) that would prevent it from applying availability effects to the associated Control. This might also come in handy outside of the keybinding scenario.

Another change would involve extension of the [Parser](https://github.com/Caliburn-Micro/Caliburn.Micro/blob/63a49070f765c126d73171722b30f4ca7471ab6d/src/Caliburn.Micro.Platform/Parser.cs#L73) that would allow message customization based on Trigger type. I don't know if this is a best approach to the problem, but I would like to see this resolved in some way.

**Comment by KasperSK** [comment]
Could you not just omit the `CanExecute` method? If the `Parser` was to know about Trigger type then it would have to be customizable for the user. I supposed the simple solution would be to add a do not validate `CanExecute` property on the `ActionMessage`, but then again why not just omit the `CanExecute` method in the first place?

**Comment by dmlewis08** [comment]
I'm running into this same issue today (2/24/2025) with Caliburn.Micro 4.0.230. I'm not seeing a relevant property on the ActionMessage to prevent the entire control from enabling and disabling with the key trigger; is there a workaround?  Sorry to revive an old issue, I didn't want to create a new one while there was an existing issue about the exact same thing.

**Comment by Filipsi** [comment]
Yeah, this was never fixed.

I just ended up using [CallMethodAction](https://github.com/microsoft/XamlBehaviorsWpf/wiki/CallMethodAction) and binding _IsEnabled_ property of the control to the guard property manually.

```xaml
<behaviors:EventTrigger EventName="KeyUp">
   <behaviors:CallMethodAction MethodName="HandleKeyUp" TargetObject="{Binding}" />
</behaviors:EventTrigger>
```

**Comment by vb2ae** [comment]
@dmlewis08 do you have an example app that shows this issue?  I can look at it

**Comment by dmlewis08** [comment]
I've put together a very simple project that demonstrates the issue:
[CaliburnMicroKeyBindingEnablingTest](https://github.com/dmlewis08/CaliburnMicroKeyBindingEnablingTest/tree/master/CaliburnMicroKeyBindingEnablingTest)

**Comment by vb2ae** [comment]
Thanks

**Comment by dmlewis08** [comment]
> I just ended up using [CallMethodAction](https://github.com/microsoft/XamlBehaviorsWpf/wiki/CallMethodAction) and binding _IsEnabled_ property of the control to the guard property manually.
> 
> <behaviors:EventTrigger EventName="KeyUp">
>    <behaviors:CallMethodAction MethodName="HandleKeyUp" TargetObject="{Binding}" />
> </behaviors:EventTrigger>

The only issue with that strategy is that it won't work for methods that return a Task for asynchronous tasks (specifically, it throws an error saying `"Could not find method named "ExecuteAsync" on object of type "ShellViewModel" that matches the expected signature`), whereas the normal Caliburn.Micro ActionMessage handles them without any problems.

> I think the user should be responsible for handling this. For example, by specifying what type of action to use in the shorthand. Something like using a `Call` keyword instead of `Action` - special behavior is then handled by the parser.
> 
> <fluent:Button Header="Go!" cal:Message.Attach="[Key Enter] = [Call MethodName]" />
> Otherwise, it's just how ActionMessage is supposed to work - you can't define a exception based on what kind of trigger behavior executed the action.

This method doesn't seem to work for attaching gesture messages to the overall window (nothing happens, there's no error message to describe).

Just for anyone interested, I've updated my example code with [another branch](https://github.com/dmlewis08/CaliburnMicroKeyBindingEnablingTest/tree/Trying-more-options) that show some more options with unexpected results, including the Message.Attach and the CallMethodAction options.

**Comment by Filipsi** [comment]
> The only issue with that strategy is that it won't work for methods that return a Task for asynchronous tasks (specifically, it throws an error saying "Could not find method named "MethodNameAsync" on object of type "ShellViewModel" that matches the expected signature), whereas the normal Caliburn.Micro ActionMessage handles them without any problems.

Yeah, you are correct. It can't be used in this case, which is limiting.

> This method doesn't seem to work for attaching gesture messages to the overall window (nothing happens, there's no error message to describe).

That is because it's a proposed solution to the problem initially described by this issue. Moreover, if you want the `Key` or `Gesture` triggers to work in the Message.Attach, you have to register them in the parser like shown in the key-binding [sample](https://github.com/Caliburn-Micro/Caliburn.Micro/blob/f6924042584c0b76c2ea21b4f6697d89c79ab1ae/samples/scenarios/Scenario.KeyBinding/Bootstrapper.cs#L21). This is not supported by the Caliburn.Micro out of the box.
